crates/polars-stream/src/nodes/merge_sorted.rs

```diff
@@ -1,126 +1,477 @@
+
+            starting_nulls: false,
+
+            left_unmerged: VecDeque::new(),
+            right_unmerged: VecDeque::new(),
         }
     }
 }
 
 impl ComputeNode for MergeSortedNode {
     fn name(&self) -> &str {
-        "in_memory_merge_sorted"
+        "merge_sorted"
     }
 
     fn initialize(&mut self, num_pipelines: usize) {
         self.num_pipelines = num_pipelines;
     }
 
     fn update_state(&mut self, recv: &mut [PortState], send: &mut [PortState]) -> PolarsResult<()> {
-        assert!(recv.len() == 2 && send.len() == 1);
+        assert_eq!(send.len(), 1);
+        assert_eq!(recv.len(), 2);
 
         // If the output doesn't want any more data, transition to being done.
-        if send[0] == PortState::Done && !matches!(self.state, MergeSortedState::Done) {
-            self.state = MergeSortedState::Done;
+        if send[0] == PortState::Done {
+            recv[0] = PortState::Done;
+            recv[1] = PortState::Done;
+
+            self.left_unmerged.clear();
+            self.right_unmerged.clear();
+            self.state = State::Piping;
+
+            return Ok(());
         }
 
-        // If the input is done, transition to being a source.
-        if let MergeSortedState::Sink { left, right } = &mut self.state {
-            if recv[0] == PortState::Done && recv[1] == PortState::Done {
-                let left_df = left.get_output()?.unwrap();
-                let right_df = right.get_output()?.unwrap();
-                let left_s = left_df.column(self.key.as_str()).unwrap();
-                let right_s = right_df.column(self.key.as_str()).unwrap();
-                let df = _merge_sorted_dfs(
-                    &left_df,
-                    &right_df,
-                    left_s.as_materialized_series(),
-                    right_s.as_materialized_series(),
-                    true,
-                )?;
-                let mut source_node = InMemorySourceNode::new(Arc::new(df), MorselSeq::default());
-                source_node.initialize(self.num_pipelines);
-                self.state = MergeSortedState::Source(source_node);
+        if recv[0] == PortState::Done || recv[1] == PortState::Done {
+            if !self.left_unmerged.is_empty() || !self.right_unmerged.is_empty() {
+                self.state = State::Flushing;
+            } else {
+                self.state = State::Piping;
             }
+        } else {
+            self.state = State::Merging;
         }
 
-        match &mut self.state {
-            MergeSortedState::Sink { left, right, .. } => {
-                left.update_state(&mut recv[0..1], &mut [])?;
-                right.update_state(&mut recv[1..2], &mut [])?;
+        match self.state {
+            State::Merging if send[0] == PortState::Blocked => {
+                recv[0] = PortState::Blocked;
+                recv[1] = PortState::Blocked;
+            },
+
+            // If one input side is blocked and we haven't buffered anything for that side, we
```

Comment:
I don't think this needs to be its own case, or needs to check `*_unmerged.is_empty()`, I think we can merge this into the above case:
```rust
let blocked = send[0] == PortState::Blocked || recv.iter().any(|p| p == PortState::Blocked);
if blocked {
    recv[0] = PortState::Blocked;
    recv[1] = PortState::Blocked;
    send[0] = PortState::Blocked;
    return;
}
```
And you can apply this logic regardless of the state you are in, simplifying the rest of the logic not having to deal with blocking.